jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/grid/headerfilter/DataGridHeaderFilter.java

```diff
@@ -0,0 +1,383 @@
+        JmixButton applyButton = uiComponents.create(JmixButton.class);
+        applyButton.setIcon(VaadinIcon.CHECK.create());
+        applyButton.setText(messages.getMessage("columnFilter.apply.text"));
+
+        applyButton.addThemeVariants(ButtonVariant.LUMO_PRIMARY);
+        applyButton.addClickListener(__ -> {
+            doApply();
+            overlay.close();
+
+            fireEvent(new ApplyEvent(this, propertyFilter, false));
+        });
+
+        if (isSmallDevice()) {
+            applyButton.getStyle().set("flex-grow", "1");
+        }
+
+        return applyButton;
+    }
+
+    protected JmixButton createCancelButton() {
+        JmixButton cancelButton = uiComponents.create(JmixButton.class);
+        cancelButton.setIcon(VaadinIcon.BAN.create());
+        cancelButton.setText(messages.getMessage("columnFilter.cancel.text"));
+
+        cancelButton.addClickListener(__ -> doCancel());
+
+        if (isSmallDevice()) {
+            cancelButton.getStyle().set("flex-grow", "1");
+        }
+
+        return cancelButton;
+    }
+
+    /**
+     * Applies the current value of the {@link PropertyFilter} to the loader
+     */
+    public void doApply() {
+        propertyFilter.getDataLoader().load();
+        appliedValue = propertyFilter.getValue();
+        appliedOperation = propertyFilter.getOperation();
+
+        filterButton.getElement().setAttribute(COLUMN_FILTER_BUTTON_ACTIVATED_ATTRIBUTE_NAME,
+                propertyFilter.getValue() != null);
+    }
+
+    @SuppressWarnings({"unchecked"})
+    protected void doCancel() {
+        propertyFilter.setValue(appliedValue);
+        propertyFilter.setOperation(appliedOperation);
+
+        overlay.close();
+    }
+
+    protected JmixButton createClearFilterButton() {
+        JmixButton clearButton = uiComponents.create(JmixButton.class);
+
+        clearButton.addThemeVariants(ButtonVariant.LUMO_ICON);
+        clearButton.setIcon(VaadinIcon.ERASER.create());
+        clearButton.addClickListener(__ -> propertyFilter.clear());
+
+        return clearButton;
+    }
+
+    protected String getOverlayPositionExpression() {
+        return "$0.$.overlay.$.overlay.style['top'] = $1.getBoundingClientRect().top + 'px';" +
+                "const sum = $1.getBoundingClientRect().left + $1.getBoundingClientRect().width " +
+                "+ $0.$.overlay.$.overlay.getBoundingClientRect().width;" +
+                "if (sum < window.innerWidth) { " +
+                "$0.$.overlay.$.overlay.style['left'] = $1.getBoundingClientRect().left + 'px'; " +
+                "} else { " +
+                "$0.$.overlay.$.overlay.style['right'] = window.innerWidth - $1.getBoundingClientRect().left " +
+                "- $1.getBoundingClientRect().width + 'px';" +
+                "}";
+    }
+
+    protected boolean isSmallDevice() {
+        // magic number from vaadin-app-layout.js
+        // '--vaadin-app-layout-touch-optimized' style property
```

Comment:
Can't we get actual `--vaadin-app-layout-touch-optimized` value using JS? 801 is a default value that is replaced with actual from `--vaadin-app-layout-touch-optimized` since  JS invocation is async

Reply:
It is impossible to determine the value of a variable on the server side at the right time.